clinicadl/utils/caps_dataset/caps_dataset_refactoring/caps_dataset.py

```diff
@@ -0,0 +1,1224 @@
+# coding: utf8
+
+import abc
+from logging import getLogger
+from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+import torch
+import torchio as tio
+import torchvision.transforms as transforms
+from clinica.utils.exceptions import ClinicaCAPSError
+from torch.utils.data import Dataset
+
+from clinicadl.prepare_data.prepare_data_utils import (
+    PATTERN_DICT,
+    TEMPLATE_DICT,
+    compute_discarded_slices,
+    compute_folder_and_file_type,
+    extract_patch_path,
+    extract_patch_tensor,
+    extract_roi_path,
+    extract_roi_tensor,
+    extract_slice_path,
+    extract_slice_tensor,
+    find_mask_path,
+)
+from clinicadl.utils.exceptions import (
+    ClinicaDLArgumentError,
+    ClinicaDLConfigurationError,
+    ClinicaDLTSVError,
+)
+
+logger = getLogger("clinicadl")
+
+
+#################################
+# Datasets loaders
+#################################
+class CapsDataset(Dataset):
+    """Abstract class for all derived CapsDatasets."""
+
+    def __init__(
+        self,
+        caps_directory: Path,
+        tsv_label: Path,
+        preprocessing_dict: Dict[str, Any],
+        transformations: Optional[Callable],
+        label_presence: bool,
+        label: str = None,
+        label_code: Dict[Any, int] = None,
+        augmentation_transformations: Optional[Callable] = None,
+    ):
+        from clinicadl.utils.preprocessing import read_preprocessing
+
+        self.caps_directory = caps_directory
+        self.caps_dict = self.create_caps_dict(caps_directory)
+        self.transformations = transformations
+        self.augmentation_transformations = augmentation_transformations
+        self.eval_mode = False
+        self.label_presence = label_presence
+        self.label = label
+        self.label_code = label_code
+        self.preprocessing_dict = read_preprocessing(preprocessing_dict)
+
+        if not hasattr(self, "elem_index"):
+            raise AttributeError(
+                "Child class of CapsDataset must set elem_index attribute."
+            )
+        if not hasattr(self, "mode"):
+            raise AttributeError("Child class of CapsDataset, must set mode attribute.")
+
+        self.df = pd.read_csv(tsv_label, sep="\t")
+        mandatory_col = {
+            "participant_id",
+            "session_id",
+        }
+
+        # check label presence and label to include it in the mandatory col
+        if self.label_presence and self.label is not None:
+            mandatory_col.add(self.label)
+
+        if not mandatory_col.issubset(set(self.df.columns.values)):
+            raise Exception(
+                f"the tsv file is not in the correct format."
+                f"Columns should include {mandatory_col}"
+            )
+        self.elem_per_image = self.num_elem_per_image()
+        self.size = self.size_elem()
+
+    @property
+    @abc.abstractmethod
+    def elem_index(self):
```

Comment:
From what I understand, but might be wrong: 

```suggestion
    def elem_index(self) -> Optional[int]:
```